File: 125. Valid Palindrome/125. Valid Palindrome.js
function isPalindrome(s) {
    // Step 1: Normalize the string
    // Convert to lowercase and remove non-alphanumeric characters
    const normalized = s.toLowerCase().replace(/[^a-z0-9]/g, '');

    // Step 2: Check for palindrome
    // Compare the normalized string with its reverse
    const reversed = normalized.split('').reverse().join('');
    return normalized === reversed;
}

// Example usage:
console.log(isPalindrome("A man, a plan, a canal: Panama")); // Output: true
console.log(isPalindrome("race a car")); // Output: false
console.log(isPalindrome(" ")); // Output: true

File: 125. Valid Palindrome/125. Valid Palindrome.rs
impl Solution {
    pub fn is_palindrome(s: String) -> bool {
        // Convert to lowercase and filter out non-alphanumeric characters
        let chars: Vec<char> = s.chars()
            .filter(|c| c.is_alphanumeric())
            .map(|c| c.to_ascii_lowercase())
            .collect();

        // Check if the string is a palindrome
        chars.iter().eq(chars.iter().rev())
    }
}

// The struct Solution is already defined in the problem framework
// No need to define it here